### ISSUE TITLE
Help with converting emmeans code to marginaleffects (multinom model)

### ISSUE THREAD

**Issue by jgeller112** [vincentarelbundock/marginaleffects#1033]
Hi, 

Love the package! I am converting my emmeans code to marginaleffects for my students, however running into an issue with multinomial models. 

I need to get the difference between each  response category level and the baseline along with the difference between the IV here PhysActive (OR). Below is some reproducible code to get this in emmeans.  

```
library(NHANES)
nhanes_adult <- NHANES %>%
  #only use ages 18+
  filter(Age >= 18) %>%
  #select 4 vars from the full dataset
  dplyr::select(HealthGen, Education, Age, PhysActive) %>%
  # get rid of nas
  drop_na() %>%
  mutate(obs_num = 1:n())

```
```
#fit multinom model
health_m <- multinom(HealthGen ~ PhysActive + Age, 
                     data = nhanes_adult)
```

```

library(emmeans)


multi_an <- emmeans(health_m, ~ PhysActive|HealthGen)

# uses baseline as contrast of interest
# can change this to get other baselines
# use trt.vs.ctrl" #ref = newbaseline
coefs = contrast(regrid(multi_an, "log"),"trt.vs.ctrl1",  by="PhysActive")
```

```
contrast(coefs, "revpairwise", by = "contrast") %>%

```

I am not sure how to replicate this in marginaleffects with `avg_comparisons`

**Comment by vincentarelbundock** [comment]
Glad you like the package\!

I’m on my wait to teach, so I can’t figure out everything now, but hopefully these examples (minus re-griding) can help get you on the right path. Note that I use `hypothesis="reference"`, but that there are other options like "pairwise". I'm also using `datagrid(grid_type="balanced")` because a specificity of `emmeans` is that it usees balanced grids by default.

``` r
library(NHANES)
library(tidyverse)
library(nnet)
library(emmeans)
library(marginaleffects)
options(width = 10000)
nhanes_adult <- NHANES %>%
  #only use ages 18+
  filter(Age >= 18) %>%
  dplyr::select(HealthGen, Education, Age, PhysActive) %>%
  drop_na() %>%
  mutate(obs_num = 1:n())

health_m <- multinom(HealthGen ~ PhysActive + Age, data = nhanes_adult, trace = FALSE)
multi_an <- emmeans(health_m, ~ PhysActive|HealthGen)

contrast(multi_an, "trt.vs.ctrl1", by = "PhysActive")
# PhysActive = No:
#  contrast          estimate      SE df t.ratio p.value
#  Vgood - Excellent   0.1753 0.00971 12  18.046  <.0001
#  Good - Excellent    0.3675 0.01115 12  32.968  <.0001
#  Fair - Excellent    0.1356 0.00927 12  14.635  <.0001
#  Poor - Excellent   -0.0236 0.00628 12  -3.753  0.0096
# 
# PhysActive = Yes:
#  contrast          estimate      SE df t.ratio p.value
#  Vgood - Excellent   0.2388 0.01213 12  19.680  <.0001
#  Good - Excellent    0.2016 0.01187 12  16.988  <.0001
#  Fair - Excellent   -0.0679 0.00842 12  -8.060  <.0001
#  Poor - Excellent   -0.1480 0.00648 12 -22.850  <.0001
# 
# P value adjustment: dunnettx method for 4 tests

avg_predictions(
    health_m,
    by = "PhysActive",
    hypothesis = "reference",
    newdata = datagrid(grid_type = "balanced")
)
# 
#                            Term Estimate Std. Error      z Pr(>|z|)     S    2.5 %  97.5 %
#  Excellent, Yes - Excellent, No   0.0861    0.00786  10.96  < 0.001  90.4  0.07071  0.1015
#  Vgood, No - Excellent, No        0.1753    0.00970  18.07  < 0.001 240.1  0.15627  0.1943
#  Vgood, Yes - Excellent, No       0.3250    0.00963  33.75  < 0.001 827.0  0.30610  0.3438
#  Good, No - Excellent, No         0.3670    0.01112  32.99  < 0.001 790.6  0.34520  0.3888
#  Good, Yes - Excellent, No        0.2874    0.00947  30.35  < 0.001 669.7  0.26881  0.3059
#  Fair, No - Excellent, No         0.1357    0.00925  14.68  < 0.001 159.6  0.11760  0.1538
#  Fair, Yes - Excellent, No        0.0182    0.00670   2.72  0.00654   7.3  0.00509  0.0314
#  Poor, No - Excellent, No        -0.0232    0.00627  -3.70  < 0.001  12.2 -0.03553 -0.0109
#  Poor, Yes - Excellent, No       -0.0619    0.00485 -12.78  < 0.001 121.8 -0.07142 -0.0524
# 
# Columns: term, estimate, std.error, statistic, p.value, s.value, conf.low, conf.high 
# Type:  probs

coefs = contrast(multi_an, "trt.vs.ctrl1", by = "PhysActive")
contrast(coefs, "revpairwise", by = "contrast")
# contrast = Vgood - Excellent:
#  contrast1 estimate      SE df t.ratio p.value
#  Yes - No    0.0635 0.01566 12   4.057  0.0016
# 
# contrast = Good - Excellent:
#  contrast1 estimate      SE df t.ratio p.value
#  Yes - No   -0.1659 0.01641 12 -10.108  <.0001
# 
# contrast = Fair - Excellent:
#  contrast1 estimate      SE df t.ratio p.value
#  Yes - No   -0.2035 0.01261 12 -16.138  <.0001
# 
# contrast = Poor - Excellent:
#  contrast1 estimate      SE df t.ratio p.value
#  Yes - No   -0.1245 0.00905 12 -13.748  <.0001

avg_comparisons(
    health_m,
    hypothesis = "reference",
    variables = "PhysActive",
    newdata = datagrid(grid_type = "balanced")
)
# 
#               Term Estimate Std. Error      z Pr(>|z|)     S   2.5 %  97.5 %
#  Fair - Excellent   -0.2036    0.01261 -16.14   <0.001 192.3 -0.2283 -0.1789
#  Good - Excellent   -0.1657    0.01641 -10.10   <0.001  77.2 -0.1979 -0.1336
#  Poor - Excellent   -0.1248    0.00907 -13.76   <0.001 140.8 -0.1426 -0.1070
#  Vgood - Excellent   0.0636    0.01568   4.06   <0.001  14.3  0.0329  0.0943
# 
# Columns: term, estimate, std.error, statistic, p.value, s.value, conf.low, conf.high 
# Type:  probs
```

**Comment by jgeller112** [comment]
This looks great! Thank you so much!

**Comment by jgeller112** [comment]
For some reason the balanced datagrid does not work for me: 

```
Error in datagrid(grid_type = "balanced", model = list(n = c(3, 0, 5),  : 
  Assertion on 'grid_type' failed: Must be element of set {'typical','counterfactual'}, but is 'balanced'.
```

**Comment by jgeller112** [comment]
This seems to work just fine to get logits. I am a bit worried about the datagrid part. 

```
avg_comparisons(
    health_m,
    hypothesis = "reference",
    variables = "PhysActive",
    type="latent") %>%
  kable()

```

**Comment by vincentarelbundock** [comment]
Could you update to 0.18.0 from CRAN or to the latest Github version, and restart R?

I think "balanced" may be a newer feature.

**Comment by jgeller112** [comment]
Everything works wonderfully!

**Comment by jgeller112** [comment]
hmm. So when I add a variable with multiple levels things seem to break down a bit. I cant get avg_comparisons to give me what emmeans does without referencing my model. 

```
health_m <- multinom(HealthGen ~ PhysActive + Age + Education, data = nhanes_adult, trace = FALSE)
multi_an <- emmeans(health_m, ~ Education|HealthGen)

coefs = contrast(regrid(multi_an, "log"),"trt.vs.ctrl1", by="Education")
update(coefs, by = "contrast")
```

```
avg_comparisons(health_m, 
hypothesis = "reference",
variables  = "Education",
type="latent")
```

**Comment by jgeller112** [comment]
```
avg_comparisons(health_m, 
hypothesis = "pairwise",
variables  = "Education",
type="latent")

```

This is just a big mess compared to nice looking results from emmeans

**Comment by vincentarelbundock** [comment]
Hmm, I could probably implement something to make these comparisons easier. But I don't really understand what the `emmeans` command does under the hood. Would you be able to show me --- using only Base R functions --- how `emmeans` arrives at one specific estimate, such as:

```r
contrast = Vgood - Excellent:
 Education      estimate     SE df t.ratio p.value
 8th Grade         0.456 0.2586 28   1.765  0.2970
```

**Comment by jgeller112** [comment]
I'm essentially looking for similar behavior to the 2 level condition where I get the log odds for the difference between two levels (8th grade and 9th grade). I'm not sure why there is different behavior in avg_compar.

I had a convo with Russ a year or so about this (will try and find).

**Comment by jgeller112** [comment]
I am thinking this issue is related to: https://github.com/vincentarelbundock/marginaleffects/issues/1024.

**Comment by jgeller112** [comment]
Additional issues: 

 the hypothesis = "reference" does not take into account releveling done on the factor (only takes the first category alphabetically)

**Comment by vincentarelbundock** [comment]
@jgeller112 and @mattansb, I just pushed a powerful new feature to `main` on Github: the `hypothesis` argument now accepts arbitrary aggregation function. Users can define their own fully customized aggregation function.

*Most importantly*, we can add new string shortcuts like “revpairwise” to apply the common use-cases that you both have been requesting. What I would need is a clear definition or “typology” of aggregations and comparisons you might care about.

### Example: multinomial model

``` r
library(NHANES)
library(tidyverse)
library(nnet)
library(emmeans)
library(marginaleffects)
options(width = 10000)

nhanes_adult <- NHANES %>%
  filter(Age >= 18) %>%
  dplyr::select(HealthGen, Education, Age, PhysActive) %>%
  drop_na() %>%
  mutate(obs_num = 1:n())

health_m <- multinom(HealthGen ~ PhysActive + Age + Education, data = nhanes_adult, trace = FALSE)
```

### Marginal means

Now, we compute equivalent marginal means with `emmeans` and `marginaleffects`:

``` r
mm_e <- emmeans(health_m, ~ Education | HealthGen)

mm_m <- avg_predictions(health_m,
    by = "Education",
    newdata = datagrid(grid_type = "balanced"),
    type = "probs")
```

### Contrasts between treatment groups

First, we compute contrasts between treatment groups with `emmeans`:

``` r
coefs = contrast(mm_e, "trt.vs.ctrl1", by = "Education")
update(coefs, by = "contrast")
# contrast = Vgood - Excellent:
#  Education      estimate      SE df t.ratio p.value
#  8th Grade        0.0481 0.02666 28   1.804  0.2789
#  9 - 11th Grade   0.1616 0.01963 28   8.231  <.0001
#  High School      0.1936 0.01585 28  12.215  <.0001
#  Some College     0.2378 0.01341 28  17.742  <.0001
#  College Grad     0.2440 0.01691 28  14.431  <.0001
# 
# contrast = Good - Excellent:
#  Education      estimate      SE df t.ratio p.value
#  8th Grade        0.2860 0.03400 28   8.412  <.0001
#  9 - 11th Grade   0.3666 0.02236 28  16.390  <.0001
#  High School      0.3503 0.01728 28  20.266  <.0001
#  Some College     0.3395 0.01414 28  24.001  <.0001
#  College Grad     0.1634 0.01635 28   9.997  <.0001
# 
# contrast = Fair - Excellent:
#  Education      estimate      SE df t.ratio p.value
#  8th Grade        0.2743 0.03277 28   8.370  <.0001
#  9 - 11th Grade   0.1705 0.01884 28   9.046  <.0001
#  High School      0.0860 0.01348 28   6.377  <.0001
#  Some College     0.0199 0.01002 28   1.983  0.2059
#  College Grad    -0.1035 0.01135 28  -9.119  <.0001
# 
# contrast = Poor - Excellent:
#  Education      estimate      SE df t.ratio p.value
#  8th Grade       -0.0242 0.01966 28  -1.230  0.6004
#  9 - 11th Grade  -0.0315 0.01137 28  -2.773  0.0410
#  High School     -0.0653 0.00872 28  -7.493  <.0001
#  Some College    -0.0681 0.00745 28  -9.143  <.0001
#  College Grad    -0.1560 0.00947 28 -16.477  <.0001
# 
# Results are averaged over the levels of: PhysActive 
# P value adjustment: dunnettx method for 5 tests
```

Now, we do the same with `marginaleffects`. The trick is to define a function that operates on the full data frame produced by the `avg_predictions()` call above. This function can do any kind of aggregation: rations, differences, cross-groups, cross-levels, whatever. It only needs to return a data frame with `term` and `estimate` columns.

Obviously, this is annoying to specify for users. The killer feature would be to create a set of pre-defined aggregation function like this one, and to tie them to shortcut keywords like the currently existing `"pairwise"`, or `"sequence"`. Maybe we could have `"group_reference"` or something like that.

``` r
group_reference <- function(x) {
  x |>
    mutate(
      term = sprintf("%s - %s", group, group[1]),
      estimate = estimate - estimate[1],
      .by = "Education") |>
    filter(term != "Excellent - Excellent") |>
    select(-group)
}

avg_predictions(health_m,
  hypothesis = group_reference,
  by = "Education",
  newdata = datagrid(grid_type = "balanced"),
  type = "probs")
# 
#               Term      Education Estimate Std. Error      z Pr(>|z|)     S     2.5 %   97.5 %
#  Vgood - Excellent 8th Grade        0.0481    0.02666   1.81   0.0709   3.8 -0.004108  0.10038
#  Vgood - Excellent 9 - 11th Grade   0.1617    0.01964   8.23   <0.001  52.3  0.123193  0.20018
#  Vgood - Excellent High School      0.1937    0.01586  12.22   <0.001 111.6  0.162652  0.22480
#  Vgood - Excellent Some College     0.2380    0.01342  17.73   <0.001 231.3  0.211666  0.26428
#  Vgood - Excellent College Grad     0.2442    0.01691  14.44   <0.001 154.7  0.211050  0.27732
#  Good - Excellent  8th Grade        0.2856    0.03397   8.41   <0.001  54.4  0.219060  0.35221
#  Good - Excellent  9 - 11th Grade   0.3662    0.02236  16.38   <0.001 197.8  0.322412  0.41007
#  Good - Excellent  High School      0.3500    0.01728  20.25   <0.001 300.4  0.316118  0.38387
#  Good - Excellent  Some College     0.3392    0.01416  23.96   <0.001 419.0  0.311416  0.36690
#  Good - Excellent  College Grad     0.1632    0.01634   9.99   <0.001  75.6  0.131181  0.19523
#  Fair - Excellent  8th Grade        0.2743    0.03275   8.38   <0.001  54.0  0.210091  0.33846
#  Fair - Excellent  9 - 11th Grade   0.1705    0.01885   9.05   <0.001  62.5  0.133541  0.20742
#  Fair - Excellent  High School      0.0860    0.01348   6.38   <0.001  32.4  0.059571  0.11242
#  Fair - Excellent  Some College     0.0199    0.01003   1.98   0.0475   4.4  0.000223  0.03955
#  Fair - Excellent  College Grad    -0.1035    0.01135  -9.12   <0.001  63.5 -0.125756 -0.08127
#  Poor - Excellent  8th Grade       -0.0238    0.01968  -1.21   0.2267   2.1 -0.062363  0.01478
#  Poor - Excellent  9 - 11th Grade  -0.0313    0.01139  -2.75   0.0060   7.4 -0.053618 -0.00897
#  Poor - Excellent  High School     -0.0652    0.00873  -7.47   <0.001  43.5 -0.082312 -0.04808
#  Poor - Excellent  Some College    -0.0680    0.00747  -9.10   <0.001  63.3 -0.082606 -0.05333
#  Poor - Excellent  College Grad    -0.1559    0.00947 -16.45   <0.001 199.7 -0.174435 -0.13730
# 
# Columns: term, Education, estimate, std.error, statistic, p.value, s.value, conf.low, conf.high 
# Type:  probs
```

**Comment by mattansb** [comment]
Okay, this is awesome!

Can you explain how this magic works? How do you get SEs from that two column df???

I'll cook up a list of aggregators, but as a start - those that are available in comparisons() would probably be good.

**Comment by jgeller112** [comment]
Awesome! How difficult would it be to get the differences between each level of education and each term. 

In emmeans you need to create an additional contrast:

```
contrast(coefs, "revpairwise", by = "contrast")
```

**Comment by vincentarelbundock** [comment]
haha, there's isn't much magic. What happened is that very early on in development I stumbled onto a *very* flexible software design to apply the delta method, and I've just been reaping winnings from that ever since. There's a single engine behind the scenes to compute standard errors for all the quantities in the package --- including the ones from that 2 column data frame. It follows the broad outlines here: https://marginaleffects.com/vignettes/uncertainty.html

I'd love to get some aggregator contributions. But note that I don't import `dplyr`. Base R works, `data.table` too. But if you write something in `dplyr` I can always translate them.

**Comment by vincentarelbundock** [comment]
> Awesome! How difficult would it be to get the differences between each level of education and each term.

Very easy. 

First, look at the output of this function:

```r
mm_m <- avg_predictions(health_m,
    by = "Education",
    newdata = datagrid(grid_type = "balanced"),
    type = "probs")
```

Second, write a function with `dplyr` to compute the quantity you care about based on the data frame in `mm_m`. This can be any transformation at all with `group_by()`, `mutate()`, `filter()`, or any base R command. Whatever you want.

Third, feed that function to the `hypothesis` argument.

**Comment by jgeller112** [comment]
I'm using this in my graduate psych class and I'm looking for something more out of the box (like emmeans). I might have to resort back to emmeans for simplicity's sake.

**Comment by vincentarelbundock** [comment]
@jgeller112 

What I'm saying is that it is now trivial for developers (Mattan and me) to implement out of the box shortcuts.

Can you give me a detailed list of comparisons your students need to conduct?

I could probably implement those as needed.

**Comment by jgeller112** [comment]
I just want the log odds for the difference between the levels of Education (8th grade - 9th. grade). It works for the two level example: 

```
avg_comparisons(
    health_m,
    hypothesis = "reference",
    variables = "PhysActive",
    type="latent")
```
There is also the issue where `reference` uses the first level alphabetically and not the reference level for the actual categorical variable (i.e., if I want poor as my reference instead). I am probably missing how that is done in marginaleffects.

**Comment by mattansb** [comment]
@vincentarelbundock I've tried using the new interface, and I have some suggestions:

1. the input to the "hypothesis" function should contain all the columns (currently I'm only seeing `rowid` and `estimate`, which is not enough for my uses).
2. the output should allow to have arbitrary columns, as long as it has estimate and term

This way, we could do things like conditional contrasts like so:

```R
library(tidyverse)
library(marginaleffects)


mod <- glm(am ~ factor(cyl) + hp, 
           family = binomial("logit"),
           data = mtcars)

w <- function(df) {
  df |> 
    group_by(hp) |> 
    summarise(
      term = "low vs high",
      estimate = estimate[cyl == 4] - estimate[cyl == 8]
    ) |> 
    ungroup() |> 
    mutate(
      term = sprintf("%s [%f]", term, hp)
    )
}

predictions(mod, newdata = datagrid(cyl = unique, hp = c(100, 110)),
            hypothesis = w)



```

**Comment by vincentarelbundock** [comment]
@mattansb, I can’t give you all columns in a single data frame because there would be a big performance penalty in more “normal” use cases.

I *can* give you separate `x` and `newdata` frames with a `rowid` column which you can easily merge. (Already available in 
 Github `main`)

Note that this example retains an arbitrary output column called `letters`.

``` r
options(width = 10000)
library(tidyverse)
library(marginaleffects)

mod <- glm(am ~ factor(cyl) + hp, family = binomial("logit"), data = mtcars)

w <- function(x, newdata) {
  x |> 
    left_join(newdata, "rowid") |>
    summarise(
      term = "low vs high",
      estimate = estimate[cyl == 4] - estimate[cyl == 8],
      .by = "hp") |> 
    mutate(
      term = sprintf("%s [%s]", term, hp),
      letters = sample(letters, n(), replace = TRUE))
}

p <- predictions(mod, 
    type = "response",
    newdata = datagrid(cyl = unique, hp = c(100, 110)),
    hypothesis = w)

p |> print("data.frame")
#                term  estimate std.error statistic      p.value  s.value  conf.low conf.high  hp letters
# 1 low vs high [100] 0.8532370 0.1067568  7.992345 1.323963e-15 49.42406 0.6439976  1.062476 100       m
# 2 low vs high [110] 0.8939935 0.0924571  9.669279 4.072383e-22 71.05654 0.7127809  1.075206 110       q
```

**Comment by mattansb** [comment]
Oh yeah, that's more like it!

```R
library(tidyverse)

library(emmeans)
library(marginaleffects)


mod <- glm(am ~ factor(cyl) + hp, 
           family = binomial("logit"),
           data = mtcars)

w <- data.frame(
  "{4,6} vs 8" = c(-.5, -.5, 1),
  "4 vs 6" = c(-1, 1, 0), 
  check.names = FALSE
)

emmeans(mod, ~ cyl + hp, at = list(hp = c(100, 110))) |> 
  contrast(method = w, by = "hp") |> regrid()
#> hp = 100:
#>  contrast   odds.ratio      SE  df z.ratio p.value
#>  {4,6} vs 8   0.000741 0.00223 Inf   0.332  0.7398
#>  4 vs 6       0.056487 0.07681 Inf   0.735  0.4621
#> 
#> hp = 110:
#>  contrast   odds.ratio      SE  df z.ratio p.value
#>  {4,6} vs 8   0.000741 0.00223 Inf   0.332  0.7398
#>  4 vs 6       0.056487 0.07681 Inf   0.735  0.4621


fw <- function(x, newdata, by, draws) {
  browser()
  x |> 
    left_join(newdata, by = "rowid") |> 
    group_by(hp) |> 
    reframe(
      term = colnames(w),
      estimate = as.vector(estimate %*% as.matrix(w)),
    ) |> 
    ungroup() |> 
    mutate(
      term = sprintf("%s [hp: %s]", term, hp),
      estimate = exp(estimate)
    )
}

predictions(mod, newdata = datagrid(cyl = unique, hp = c(100, 110)),
            hypothesis = fw,
            type = "link")
#> 
#>                  Term  hp Estimate Std. Error     z Pr(>|z|)   S    2.5 %  97.5 %
#>  {4,6} vs 8 [hp: 100] 100 0.000741    0.00223 0.332    0.740 0.4 -0.00363 0.00511
#>  4 vs 6 [hp: 100]     100 0.056487    0.07681 0.735    0.462 1.1 -0.09406 0.20703
#>  {4,6} vs 8 [hp: 110] 110 0.000741    0.00223 0.332    0.740 0.4 -0.00363 0.00511
#>  4 vs 6 [hp: 110]     110 0.056487    0.07681 0.735    0.462 1.1 -0.09406 0.20703
#> 
#> Columns: term, estimate, std.error, statistic, p.value, s.value, conf.low, conf.high, hp 
#> Type:  link 
```

As for the default schemes, going by `help("emmc-functions")`:

- pairwise / revpairwise
- poly
- consec
- trt.vs.ctrl (treatment)
- eff (deviation from mean)
- del.eff (deviation from mean excluding level)

**Comment by jgeller112** [comment]
> Additional issues:
> 
> the `hypothesis = "reference"` argument does not take into account releveling done on the factor (only takes the first category alphabetically)

Is this a bug or am I not setting an argument correctly?

**Comment by vincentarelbundock** [comment]
> Is this a bug or am I not setting an argument correctly?

@jgeller112 

This is just not a use-case I had ever thought about, so there's nothing implemented to handle this gracefully. I'm not sure I'd call this a "bug"; more a lack of feature.

I do agree with you that this is important and we need better & easier handling. This thread is getting unwieldy, so I opened a separate issue to track this specific feature request and will tag you there once I have a solution for you:

https://github.com/vincentarelbundock/marginaleffects/issues/1033

I will also ping you in a separate thread once we've settled on a user-friendly interface to make the kinds of outcome-level comparisons you were requesting in the first place. I think we are very close to a clean and easy solution. All the mechanics are there, but I want to make sure we nail the user interface before releasing code.

**Comment by vincentarelbundock** [comment]
The development version of `marginaleffects` on Github is not much more flexible in accepting custom functions in the `hypothesis` argument of all its functions. This is an extremely powerful strategy, but it can be tedious to create appropriate labels or apply functions by subgroups.

Thus, I have created a `specify_hypothesis()` function. This is a helper that returns a function which can then be passed to the `hypothesis` argument for complex aggregation or hypothesis testing. This is powerful, because it not only allows differences, but also any user-supplied function to compare estimates.

Admittedly, this still feels like a power user feature, but I like it because it's sooooo flexible. I'm keen to get feedback and suggestions on how to make this easier more intuitive. At least the machinery is all there now.

Tutorial here: https://marginaleffects.com/vignettes/hypothesis.html#functions

I recommend reading the "Functions" section from the top before getting to `specify_hypothesis()`.

**Comment by mattansb** [comment]
This looks awesome @vincentarelbundock ! 
I will try to use this next time I have complex contrasts in a project. Thanks!

**Comment by vincentarelbundock** [comment]
In retrospect, I concluded that specify_hypothesis() is way too complicated. It was a mistake to export that function, and it will be deprecated.

Instead, I have implemented grouped hypothesis tests using an interface similar (but not identical) to the formula interface in `emmeans`.

I think it's pretty cool. Leaving link to the documentation here in case someone finds this issue: https://marginaleffects.com/vignettes/hypothesis.html#formulas-group-wise
